streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -5,128 +5,133 @@ import zio._
-  val push: ZManaged[R, Nothing, ZSink.Push[R, E, I, Z]]
+abstract class ZSink[-R, +E, -I, +L, +Z] private (
+  val push: ZManaged[R, Nothing, ZSink.Push[R, E, I, L, Z]]
 ) { self =>
   import ZSink.Push
 
   /**
    * Operator alias for [[race]].
    */
-  final def |[R1 <: R, E1 >: E, A0, I1 <: I, Z1 >: Z](
-    that: ZSink[R1, E1, I1, Z1]
-  ): ZSink[R1, E1, I1, Z1] =
+  final def |[R1 <: R, E1 >: E, A0, I1 <: I, L1 >: L, Z1 >: Z](
+    that: ZSink[R1, E1, I1, L1, Z1]
+  ): ZSink[R1, E1, I1, L1, Z1] =
     self.race(that)
 
   /**
    * Operator alias for [[zip]].
    */
-  final def <*>[R1 <: R, E1 >: E, I1 <: I, Z1, Z2](
-    that: ZSink[R1, E1, I1, Z1]
-  ): ZSink[R1, E1, I1, (Z, Z1)] =
+  final def <*>[R1 <: R, E1 >: E, I1 <: I, L1 >: L, Z1, Z2](
+    that: ZSink[R1, E1, I1, L1, Z1]
+  )(implicit ev: L <:< I1): ZSink[R1, E1, I1, L1, (Z, Z1)] =
     zip(that)
 
   /**
    * Operator alias for [[zipPar]].
    */
-  final def <&>[R1 <: R, E1 >: E, I1 <: I, Z1](
-    that: ZSink[R1, E1, I1, Z1]
-  ): ZSink[R1, E1, I1, (Z, Z1)] =
+  final def <&>[R1 <: R, E1 >: E, I1 <: I, L1 >: L, Z1](
+    that: ZSink[R1, E1, I1, L1, Z1]
+  ): ZSink[R1, E1, I1, L1, (Z, Z1)] =
     self.zipPar(that)
 
   /**
    * Operator alias for [[zipRight]].
    */
-  final def *>[R1 <: R, E1 >: E, I1 <: I, Z1, Z2](
-    that: ZSink[R1, E1, I1, Z1]
-  ): ZSink[R1, E1, I1, Z1] =
+  final def *>[R1 <: R, E1 >: E, I1 <: I, L1 >: L, Z1, Z2](
+    that: ZSink[R1, E1, I1, L1, Z1]
+  )(implicit ev: L <:< I1): ZSink[R1, E1, I1, L1, Z1] =
     zipRight(that)
 
   /**
    * Operator alias for [[zipParRight]].
    */
-  final def &>[R1 <: R, E1 >: E, I1 <: I, Z1](that: ZSink[R1, E1, I1, Z1]): ZSink[R1, E1, I1, Z1] =
+  final def &>[R1 <: R, E1 >: E, I1 <: I, L1 >: L, Z1](that: ZSink[R1, E1, I1, L1, Z1]): ZSink[R1, E1, I1, L1, Z1] =
     self.zipParRight(that)
 
   /**
    * Operator alias for [[zipLeft]].
    */
-  final def <*[R1 <: R, E1 >: E, I1 <: I, Z1, Z2](
-    that: ZSink[R1, E1, I1, Z1]
-  ): ZSink[R1, E1, I1, Z] =
+  final def <*[R1 <: R, E1 >: E, I1 <: I, L1 >: L, Z1, Z2](
+    that: ZSink[R1, E1, I1, L1, Z1]
+  )(implicit ev: L <:< I1): ZSink[R1, E1, I1, L1, Z] =
     zipLeft(that)
 
   /**
    * Operator alias for [[zipParLeft]].
    */
-  final def <&[R1 <: R, E1 >: E, I1 <: I](that: ZSink[R1, E1, I1, Any]): ZSink[R1, E1, I1, Z] =
+  final def <&[R1 <: R, E1 >: E, I1 <: I, L1 >: L](that: ZSink[R1, E1, I1, L1, Any]): ZSink[R1, E1, I1, L1, Z] =
     self.zipParLeft(that)
 
   /**
    * Replaces this sink's result with the provided value.
    */
-  def as[Z2](z: => Z2): ZSink[R, E, I, Z2] =
+  def as[Z2](z: => Z2): ZSink[R, E, I, L, Z2] =
     map(_ => z)
 
   /**
    * Repeatedly runs the sink for as long as its results satisfy
    * the predicate `p`. The sink's results will be accumulated
    * using the stepping function `f`.
    */
-  def collectAllWhileWith[S](z: S)(p: Z => Boolean)(f: (S, Z) => S): ZSink[R, E, I, S] =
+  //TODO:
```

Comment:
I hate TODOs and commented code. Let's get rid of this entirely until we bring it back in a follow-up.